src/types.jl

```diff
@@ -26,7 +26,7 @@ DataFormat `fmt`.  For example, `File{fmtpng}(filename)` would indicate a PNG
 file.
 """
 struct File{F<:DataFormat} <: Formatted{F}
-    filename::String
+    filename
```

Comment:
```
julia> struct Foo
           x
       end

julia> struct Bar
           x::String
       end

julia> s = "foo"
"foo"

julia> @benchmark Foo(s)
BenchmarkTools.Trial:
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     6.718 ns (0.00% GC)
  median time:      8.086 ns (0.00% GC)
  mean time:        8.683 ns (1.64% GC)
  maximum time:     529.242 ns (98.16% GC)
  --------------
  samples:          10000
  evals/sample:     999

julia> @benchmark Bar(s)
BenchmarkTools.Trial:
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     6.945 ns (0.00% GC)
  median time:      7.543 ns (0.00% GC)
  mean time:        8.746 ns (1.61% GC)
  maximum time:     494.128 ns (98.33% GC)
  --------------
  samples:          10000
  evals/sample:     999
```

Reply:
Should be fine. I wouldn't mind if @timholy took a look this to ensure that loosening the constraints won't cause any issues

Reply:
The problem isn't construction, it's usage. Check for performance regression in code that accesses the `filename` field.

That said, generally when doing IO these things are irrelevant compared to the actual IO. There is a workaround,
```julia
struct File{F<:DataFormat,FT} <: Formatted{F}
    filename::FT
end
```
but most likely that would just add annoying compile-time latency. So I can support this change.

Reply:
Access time also seem comparable.

```
julia> @benchmark foo.x
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     21.980 ns (0.00% GC)
  median time:      22.613 ns (0.00% GC)
  mean time:        23.680 ns (0.00% GC)
  maximum time:     163.198 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     995

julia> @benchmark bar.x
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     22.102 ns (0.00% GC)
  median time:      22.564 ns (0.00% GC)
  mean time:        23.706 ns (0.00% GC)
  maximum time:     120.588 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     995
```

I'm not sure if this would introduce type instabilities for more complex functions though.

Yes, constructing a parameterized type seems more expensive (though I just used `@time`).

```
julia> @time struct Foo
           x
       end
  0.000437 seconds (137 allocations: 6.547 KiB)

julia> @time struct Bar
           x::String
       end
  0.000703 seconds (167 allocations: 8.300 KiB)

julia> @time struct Baz{T}
           x::T
       end
  0.001866 seconds (674 allocations: 29.738 KiB)
```